common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+/// network requests, this may represent a request or a byte.  `Tokens` are the counters for the
+/// rate limiting, and when there are no `tokens` left in a `bucket`, the `key` is throttled.
+///
+/// ## Bucket
+/// A `bucket` is the tracker of the number of `tokens`.  It has a `bucket size`, and any additional
+/// tokens added to it will "spill" out of the `bucket`.  The `buckets` are filled at an `interval`
+/// with a given `fill rate`.
+///
+/// ## Interval
+/// The `interval` at which we refill *all* of the `buckets` in the token bucket filter. Configured
+/// across the whole token bucket filter.
+///
+/// ## Fill Rate
+/// The rate at which we fill a `bucket` with tokens. Configured per bucket.
+///
+/// ## Bucket Size
+/// Maximum size of a bucket.  A bucket saturates at this size.  Configured per bucket.
+///
+/// # Features
+/// ## Keys
+/// The token bucket takes any key as long as it's hashable.  This should allow it to apply to
+/// many applications that need rate limiters.
+///
+/// ## Bucket sizes and Rates
+/// ### Defaults
+/// There are defaults for bucket size and fill rate, which will apply to unknown keys.
+///
+/// ### Per key overrides
+/// A framework was put in for allowing to make custom bucket sizes and custom fill rates
+/// There is a default bucket size, and a default fill rate, but has the framework to have custom
+/// made buckets and fill rates for specific keys that we want to be different.
+///
+/// ### Refill Interval
+/// Buckets are refilled automatically at an interval.  To do this synchronously, it calculates the
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+/// TODO: Make thread safe interface
+/// TODO: Implement an eviction policy based on inactivity (LRU)?
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    buckets: HashMap<Key, Bucket>,
+    last_refresh_time: Instant,
+    default_bucket_size: u64,
+    default_fill_rate: u64,
+    interval: Duration,
+}
+
+const ONE_SECOND: Duration = Duration::from_secs(1);
+
+#[allow(dead_code)]
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub(crate) fn new(
+        default_bucket_size: u64,
+        default_fill_rate: u64,
+        interval: Duration,
+    ) -> Self {
+        // A safeguard to ensure we don't spend all of our time refilling throttle buckets
+        assert!(
+            interval >= ONE_SECOND,
+            "Rate limiter cannot be used at less than 1 second granularity"
+        );
+
+        Self {
+            buckets: HashMap::new(),
+            last_refresh_time: Instant::now(),
+            default_bucket_size,
+            default_fill_rate,
+            interval,
+        }
+    }
+
+    /// Adds a bucket with the given configuration, defaultted to filled
+    fn ensure_bucket(&mut self, key: Key, bucket_size: u64, fill_rate: u64) {
```

Comment:
can probably remove this

Reply:
I separated this out for legibility, but yes it could be removed.

Reply:
Ah, I remember why.

So, in the event we add the builder to make custom throttles for specific keys, we could use this function accordingly to add buckets